### PR TITLE
Ampers: Angela & Phoebe

### PR DESCRIPTION
**Pull request by AngelaPoland**
# Scrabble
Congratulations! You're submitting your assignment.

## Comprehension Questions

Feature | Feedback
--- | ---
What are the advantages of using git when collaboratively working on one code base? | 

It allowed us to work on the code on our own computers and view the other person's changes.

Why do you think we chose to make the `score` method in the `Scoring` class a **class method** instead of an **instance method**? |

Any player instance should be able to use the score method per word and get the highest scoring word per game instead of creating a whole new score instance every time they played a word (which would call the class.)

Describe an example of a test you wrote for a _nominal case_. |

We were able to play a word "dragon" for play(word) method and get the correct (and upcased) word returned.

Describe an example of a test you wrote for an _edge case_. |

We definitely could have written more edge case tests. We have only one where we set a default name for the player class incase no name was given.

Does your code use any methods from the `Enumerable` mixin? If so, where and why was it helpful? |

We used map in Scrabble::Player.play(word) method to get the sum of integers from the plays array.

What was one method you and your pair used to debug code? 

We used pry in both our tests and in the files quite a bit.

AP says " I have not really used pry prior to this and am a lot more comfortable with it now."

Is there a specific piece of code you'd like feedback on? |

Where could we use more enumerable methods? 
Also, Is it possible to use one classes’ instance variable as the argument in one of the other classes’ methods? We didn't do this in the end bc it may not be possible but we'd love to know if you CAN and when would that be appropriate. (For instance, we wanted to put the TileBag instance of tile_bag into the argument for the Player's draw_tiles method.

What are two discussion points that you and your pair discussed when giving/receiving feedback from each other that you would be willing to share? |

Before we started any coding, we discussed our different comfort levels of our coding knowledge which set the tone for how we approached pair programming. Turns out, we didn't do the literal, driver/navigator dynamic in terms of sharing a computer or typing but we both feel like we equally contributed to the tasks. We dictated every next step one at a time, whiteboarded ideas of how we wanted to approach it, then Angela did the typing but we both went back and forth with our thinking through the problems. We did it this way because I, Angela, had expressed that I learn better if I am literally typing things out. Phoebe was very patient with letting me think through solutions without interjecting her own solutions and only once i had something did we both discuss the pros and cons of each together. 
And for me, Phoebe, I am able to learn/process by watching and talking through code. It was also beneficial for me to learn how to explain myself better. Talking out loud and explaining things in different ways changed how I thought about various problems. 

Since we have time tomorrow, we'll still go back over and possibly refactor more and add more test edge cases as well as adjusting our tile_bag from a hash to an array (or array of arrays) for better probabilities of the players to not get all Es (as an example.)

### CODE REVIEW

File: specs/player_spec.rb [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+
+        kamaji = Scrabble::Player.new("Kamaji")
+        bag = Scrabble::TileBag.new
+        kamaji.draw_tiles(bag)
+
+        kamaji.play("cat")
+
+        kamaji.tiles
+
+        binding.pry
```
Comment (CheezItMan) [inline specs/player_spec.rb:152]: Please leave `binding.pry` out of your project submissions.

File: lib/scoring.rb [hunk 2]
```diff
@@ -1,9 +1,116 @@
...
+          if score_array[index] == max_value
+            #if the value at the index matches max_value it will be stored in the tied_words array
+            tied_words << array_of_words[index]
+          end
+        end
+
+        smallest_word = tied_words[0]
+
+        tied_words.each do |word|
+          if word.length < smallest_word.length
```
Comment (CheezItMan) [inline lib/scoring.rb:91]: What about 7-letter words that are tied!

File: specs/scoring_spec.rb [hunk 3]
```diff
@@ -42,21 +42,34 @@
...
     end
 
     it 'returns the highest word if there are two words' do
+
+      Scrabble::Scoring.highest_score_from(["cat", "turtle"]).must_equal "TURTLE"
     end
 
     it 'if tied, prefer a word with 7 letters' do
+
+      Scrabble::Scoring.highest_score_from(["dealer", "stealer"]).must_equal "STEALER"
```
Comment (CheezItMan) [inline specs/scoring_spec.rb:61]: These two words are **not** tied.
Try it with:
```ruby
      # both score 58
      Scrabble::Scoring.highest_score_from(["ZZZZZJ", "AAAAAAD"]).must_equal "AAAAAAD"
      Scrabble::Scoring.highest_score_from(["AAAAAAD", "ZZZZZJ"]).must_equal "AAAAAAD"
```

File: lib/player.rb [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+      @tiles = []
+
+
+    end
+
+    def play(word) #checks if won and updates current word score
+
+      if @total_score > 100
+        @won = true
+        puts "Yay! You won and left the spirit world!"
```
Comment (CheezItMan) [inline lib/player.rb:28]: These methods shouldn't be doing `puts`

File: lib/player.rb [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+      @total_score = 0
+      @tiles = []
+
+
+    end
+
+    def play(word) #checks if won and updates current word score
+
+      if @total_score > 100
+        @won = true
```
Comment (CheezItMan) [inline lib/player.rb:27]: This means I have to **try** to `play` a word **after** I've won to set `@won` to true.

File: specs/player_spec.rb [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+  end
+
+  describe "plays" do
+
+    it 'returns an array of words' do
+      haku = Scrabble::Player.new("Haku")
+      haku.plays = ["Haku", "dust bunnies"]
+      haku.plays[0].must_be_kind_of String
+      haku.plays[1].must_be_kind_of String
+      haku.plays[0].wont_be_kind_of Integer
```
Comment (CheezItMan) [inline specs/player_spec.rb:29]: You probably don't need to verify that it's not an `Integer`

File: specs/player_spec.rb [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+
+    it 'returns false if player has already won' do
+
+      @first_player.total_score.wont_be:<,0
+
+      @first_player.play("radish")
+      @first_player.total_score.must_be:>,0
+
+
+      @first_player.total_score = 110
```
Comment (CheezItMan) [inline specs/player_spec.rb:59]: Don't "hack" your tests, you shouldn't be able to directly set the player's score.  Instead play enough words that the player has won, and then you can test.

File: lib/player.rb [hunk 8]
```diff
@@ -0,0 +1,70 @@
+require_relative '../lib/scoring'
+require_relative '../lib/tile-bag'
+require 'pry'
+
+
+module Scrabble
+
+  class Player
+
+    attr_accessor :name, :plays, :won, :total_score, :tile_bag, :tiles
```
Comment (CheezItMan) [inline lib/player.rb:10]: These should **not** be accessors.

File: specs/player_spec.rb [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+
+    it 'returns the score of the word' do #if player didn't win
+      @first_player.play("dragon").must_equal 8
+      @first_player.play("dog").must_equal 5
+      @first_player.plays.must_equal ["dragon", "dog"]
+      @first_player.total_score.must_equal 13
+    end
+  end
+
+  describe 'highest_scoring_word' do
```
Comment (CheezItMan) [inline specs/player_spec.rb:72]: What about when no words have been played?

File: specs/player_spec.rb [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+  end
+
+  describe "gets tiles" do
+    it "is a collection of tiles with a max of 7 letters" do
+
+      Scrabble::Player.new.tiles.must_be_kind_of Array
+
+      Scrabble::Player.new.tiles.length.must_be:<,8
+
+      Scrabble::Player.new.tiles.each do |letter|
```
Comment (CheezItMan) [inline specs/player_spec.rb:106]: You shouldn't need to create a new Player 3 times in this test.

File: specs/player_spec.rb [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+      #binding.pry
+
+      @a_player.tiles.length.must_equal 7
+
+      # edge case! we put an eigth tile into tiles to make sure an ArgumentError was raised.
+      #@a_player.tiles << "A"
+
+
+      proc {
+        @a_player.draw_tiles(bag)}.must_raise ArgumentError
```
Comment (CheezItMan) [inline specs/player_spec.rb:135]: This shouldn't raise an argumentError, it should just end up not drawing any times.

File: lib/player.rb [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+        new_tiles = []
+        num = 7 - tiles.length
+        # since tile_bag is from TileBag class, we can also call that classes' method of draw_tiles on it.
+        new_tiles = tile_bag.draw_tiles(num)
+
+        new_tiles.each do |letter|
+          @tiles << letter
+        end
+      else
+        raise ArgumentError.new ("You already have enough tiles, greedy.")
```
Comment (CheezItMan) [inline lib/player.rb:50]: This should end up not drawing any tiles, not halting the program with an error.

File: lib/tile-bag.rb [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+
+module Scrabble
+  class TileBag
+
+    attr_accessor :tiles_remaining, :tile_bag
+
+    def initialize
+      @tiles_remaining = 98
+
+      @tile_bag = {
```
Comment (CheezItMan) [inline lib/tile-bag.rb:14]: This is **not** a great structure for tracking the tiles.  This is because doing it this way makes it equally likely any letter will be drawn despite the fact that there are 9 "A" tiles and only 2 "B" tiles.

File: lib/tile-bag.rb [hunk 14]
```diff
@@ -0,0 +1,69 @@
+# require_relative '../lib/player'
+# require_relative '../lib/scoring'
+
+require 'pry'
+
+module Scrabble
+  class TileBag
+
+    attr_accessor :tiles_remaining, :tile_bag
```
Comment (CheezItMan) [inline lib/tile-bag.rb:9]: Again, these should **not** be `attr_accessor`

**Comment by CheezItMan** [comment]
# Scrabble
## What We're Looking For

Feature | Feedback
--- | ---
**General** |
Answered comprehension questions | Check, I'll look for opportunities for Enumerables.  Keep in mind that it's important each team member has keyboard time.
Both Teammates contributed to the codebase |  I assume so, but I only see Angela's commits.
Regular Commits with meaningful commit messages |  More commits is better and don't focus on "waves" but instead on the functionality you added, or code you wrote in commit messages.  
**Scoring class** |
`score` method |  Check
Uses appropriate data structure to store the letter score | Check
Appropriately handles edge cases | Check
`highest_score_from` method | **NOT WORKING**
Appropriately handles edge cases |  **NO** Does not consider 7-letter words in ties.
Test implementations for `highest_score_from`|  **DID NOT TEST A TIE WITH A 7-LETTER WORD**
**Player class** |  Do not create `attr_accesor` unless you really need to.  
Uses an array to keep track of words played |  Check
Uses existing code to create `highest_scoring_word` and `highest_word_score` |Check 
Returns `true` or `false` from `won?` method | **Sorta** see my notes
Tests edge cases on the `play(word)` |  See my notes.  You're hacking your way to an edge-case.  Good that you're trying however.  
**TileBag class** |
Uses a data structure to keep the set of default tiles in the bag | Not a good choice, it functions, but the probabilities are all off.  See my inline notes. 
`draw_tiles` method uses the parameter to draw a set of random tiles |  Check
`tiles_remaining` method returns a count |  Check
Tests `draw_tiles` edge cases |  You didn't test what happens when you try to draw all the tiles. 
Summary | There's a lot of room for improvement here.  Take a look at my notes and demo solution for `highest_score_from`.  Also take a look at your testing and spend a bit of time identifying edge-cases.  What are times when things could break or behave differently in an object.  You also shouldn't make everything an accessor.  
Additional note | When you submit the PR, note that we're using a markdown table.  If you need you can [see the documentation] on how the tables are made.  Don't disrupt them.  It the submission harder to read.
